File: admin/drawGraph.py
import svgwrite
from collections import deque
import json


class Cell:
    def __init__(self, name, url):
        self.name = name
        self.url = url
        self.content = []
        self.kid = set()
        self.parent = set()

    def to_dict(self):
        return {
            "name": self.name,
            "url": self.url,
            "kid": list(self.kid),    # Convert set to list for JSON serialization
            "parent": list(self.parent)  # Convert set to list
        }

    @classmethod
    def from_dict(cls, data):
        """Create a Cell object from a dictionary."""
        cell = cls(data["name"], data["url"])
        cell.kid = set(data["kid"])  # Convert list back to set
        cell.parent = set(data["parent"])  # Convert list back to set
        return cell

html_template = """
                <!DOCTYPE html>
                <html>
                <head>
                <meta charset="UTF-8">
                <title>SVG Example</title>
                <style>
                    #svg-container {{
                    width: 100%;
                    height: 100%;
                    overflow: auto;
                    }}
                </style>
                </head>
                <body>
                <div id="svg-container">
                    <svg width="10000" height="10000">
                    {}
                    </svg>
                </div>
                </body>
                </html>
                """

class DrawGraph:
    def __init__(self):
        with open("./cellLib.json", "r") as f:
            data = json.load(f)
            print(data)
            self.dag = {k: Cell.from_dict(v) for k, v in data.items()}

        for e in self.dag.values():
            print([e.name, e.parent, e.kid])

    def bfs (self, root, position):
        queue = deque([self.dag[root]])
        level = -1

        while queue:
            n = len(queue)
            level = level + 1

            for i in range(n):
                node = queue.popleft()
                position[node] = [level, i]

                for kid in node.kid:
                    kid = self.dag[kid]
                    if kid not in position:
                        queue.append(kid)
                        position[kid] = [-1, -1] 

    def drawDAG(self, root, filename):
        position = {}
        self.bfs(root, position)

        elements_num = [0 for _ in range(100)]
        for _, value in position.items():
            elements_num[value[0]] = max(elements_num[value[0]], value[1] + 1)
        elements_num_max = max(elements_num)

        for _, value in position.items():
            value[1] = (elements_num_max * 50) / elements_num[value[0]] * (value[1] + 0.5)
            value[0] = (value[0] + 1) * 300 - 200
        
        # svgwrite
        dwg = svgwrite.Drawing(filename, profile='full')
        dwg.defs.add(dwg.style('a:hover { cursor: pointer; }'))

        self.drawNode(dwg, self.dag[root], position, set([]))

        # 生成 HTML 文件
        with open(filename, "w", encoding="utf-8") as f:
            f.write(html_template.format(dwg.tostring()))

    def drawNode(self, dwg, node, position, visited):
        radius = 8

        # Draw the node
        dwg.add(
            dwg.a(href=node.url, target='_blank')
        ).add(
            dwg.circle(center=position[node], r=radius, stroke='red', fill='#FF8888', stroke_width=2)
        )

        # Draw children
        for i, child in enumerate(node.kid):
            child = self.dag[child]
            if (node, child) not in visited:
                visited.add((node, child))
                self.drawEdge(dwg,
                              position[node][0] + radius, position[node][1],
                              position[child][0] - radius, position[child][1])
                self.drawNode(dwg, child, position, visited)

        dwg.add(
            dwg.a(href=node.url, target='_blank')
        ).add(
            dwg.text(node.name, insert=(position[node][0] + radius * 1.5, position[node][1] + radius / 1.25), font_family="Times New Roman", font_size=20, fill="black")
        )

    def drawEdge(self, dwg, x1, y1, x2, y2):
        dwg.add(
            dwg.path(
                d=("M", x1, y1, "C", (x1 + x2) / 2, y1, (x1 + x2) / 2, y2, x2, y2),
                stroke="#8888FF", fill="none", stroke_width=1)
        )

File: admin/drawGraph.js
let dag = {};

// Load data (assuming it's stored in a file named `dag.json`)
fetch('./admin/cellLib.json').then(response => response.json()).then(data => {
    dag = data;
});

function drawDAG(root) {
    // init SVG
    const svgContainer = document.getElementById('svg-container');
    while (svgContainer.firstChild) {
        svgContainer.removeChild(svgContainer.firstChild);
    }

    const svg = document.createElementNS("http://www.w3.org/2000/svg", "svg");
    svg.setAttribute("width", "10000");
    svg.setAttribute("height", "10000");
    svgContainer.appendChild(svg);
    
    // if input root
    if (root === "") {
        const virtualRoot = {
            name: "root",
            kid: findRoots(),
            url: "#"
        };

        dag[virtualRoot.name] = virtualRoot;
        root = "root";
    }

    // Graph
    position = bfs(root);

    let elements_num = Array(100).fill(0);
    for (let key in position) {
        let value = position[key];
        elements_num[value[0]] = Math.max(elements_num[value[0]], value[1] + 1);
    }

    let elements_num_max = Math.max(...elements_num);
    for (let key in position) {
        let value = position[key];
        value[1] = (elements_num_max * 50) / elements_num[value[0]] * (value[1] + 0.5);
        value[0] = (value[0] + 1) * 300 - 200;
    }

    draw(svg, dag[root], position, new Set());

    if(root === "root") {
        delete dag[root];
    }
}

function findRoots() {
    const allNodes = new Set(Object.keys(dag));
    for (let nodeName in dag) {
        for (let kid of dag[nodeName].kid) {
            allNodes.delete(kid);
        }
    }
    return Array.from(allNodes);
}

function bfs(root) {
    const position = {};
    const queue = [dag[root]];
    let level = -1;

    while (queue.length) {
        const n = queue.length;
        level += 1;
        for (let i = 0; i < n; i++) {
            const node = queue.shift();
            position[node.name] = [level, i];
            node.kid.forEach(kidName => {
                const kid = dag[kidName];
                if (!(kidName in position)) {
                    queue.push(kid);
                    position[kidName] = [-1, -1];
                }
            });
        }
    }

    return position;
}

function draw(svg, node, position, visited) {
    const radius = 8;
    
    node.kid.forEach(kidName => {
        const child = dag[kidName];

        drawEdge(svg, 
            position[node.name][0] + radius, 
            position[node.name][1], 
            position[child.name][0] - radius, 
            position[child.name][1]);

        if (!visited.has(kidName)) {
            visited.add(kidName);
            draw(svg, child, position, visited);
        }
    });

    [x, y] = position[node.name];
    drawNode(svg, x, y, node.name, node.url);
}

function drawNode(svg, x, y, node_name, node_url) {
    const radius = 8;
    const circle = document.createElementNS("http://www.w3.org/2000/svg", "circle");
    circle.setAttribute("cx", x);
    circle.setAttribute("cy", y);
    circle.setAttribute("r", radius);
    circle.setAttribute("stroke", "red");
    circle.setAttribute("fill", "#FF8888");
    circle.setAttribute("stroke-width", "2");

    const text = document.createElementNS("http://www.w3.org/2000/svg", "text");
    text.setAttribute("x", x + radius * 1.5);
    text.setAttribute("y", y + radius / 1.25);
    text.setAttribute("font-family", "Times New Roman");
    text.setAttribute("font-size", "20");
    text.textContent = node_name;

    const circleLink = document.createElementNS("http://www.w3.org/2000/svg", "a");
    circleLink.setAttribute("href", "javascript:void(0)");  // You can use this to prevent default link action
    circleLink.addEventListener('click', () => drawDAG(node_name));

    const textLink = document.createElementNS("http://www.w3.org/2000/svg", "a");
    textLink.setAttribute("href", node_url);

    circleLink.appendChild(circle);
    textLink.appendChild(text);
    svg.appendChild(circleLink);
    svg.appendChild(textLink);
}

function drawEdge(svg, x1, y1, x2, y2) {
    const path = document.createElementNS("http://www.w3.org/2000/svg", "path");
    const d = `M ${x1} ${y1} C ${(x1 + x2) / 2} ${y1} ${(x1 + x2) / 2} ${y2} ${x2} ${y2}`;

    path.setAttribute("d", d);
    path.setAttribute("stroke", "#8888FF");
    path.setAttribute("fill", "none");
    path.setAttribute("stroke-width", "1");
    svg.appendChild(path);
}


